lib/elixir/test/elixir/kernel/warning_test.exs

```diff
@@ -1150,6 +1150,64 @@ defmodule Kernel.WarningTest do
     purge([Sample1, Sample1.Atom])
   end
 
+  describe "callbacks and friends are inside a protocol" do
+    test "warns when callbacks defined" do
+      message =
+        capture_err(fn ->
+          Code.eval_string(~S"""
+            defprotocol SampleWithCallbacks do
+              @spec with_specs(any(), keyword()) :: tuple()
+              def with_specs(term, options \\ [])
+
+              def without_specs(term, options \\ [])
+
+              @callback foo(term) :: {:ok, term}
+              @callback foo(term, keyword) :: {:ok, term, keyword}
+
+              @macrocallback bar(term) :: {:ok, term}
+              @macrocallback bar(term, keyword) :: {:ok, term, keyword}
+
+              @optional_callbacks [foo: 1, foo: 2]
+              @optional_callbacks [without_specs: 2]
+            end
+          """)
+        end)
+
+      assert message =~
+               "cannot define @callback foo/1 inside protocol, use def/1 to outline your protocol definition\n  nofile:1"
+
+      assert message =~
+               "cannot define @callback foo/2 inside protocol, use def/1 to outline your protocol definition\n  nofile:1"
+
+      assert message =~
+               "cannot define @macrocallback bar/1 inside protocol, use def/1 to outline your protocol definition\n  nofile:1"
+
+      assert message =~
+               "cannot define @macrocallback bar/2 inside protocol, use def/1 to outline your protocol definition\n  nofile:1"
+
+      assert message =~
+               "cannot define @optional_callbacks inside protocol, all of the protocol definitions are required\n  nofile:1"
+    after
+      purge([SampleWithCallbacks])
+    end
+
+    test "does not warn when callbacks are not defined" do
```

Comment:
We don't need this test. Otherwise all other protocols in Elixir and in the test suite would warn.

Reply:
It will be obvious to the eye due to the warnings, but "make test" will pass
I have added an element to the macrocallback list and I got no so many warnings, and the test pass

Reply:
Anyway, I will delete it.